// @client/graphics/Mesh.ts
import Matrix4x4 from "typed-math/Matrix4x4";
import Vector3 from "typed-math/Vector3";
import Orientation from "typed-math/Orientation";
import Engine from "@client/Engine";
import Geometry from "./Geometry";

export default class Mesh {
	position: Vector3 = new Vector3();
	scale: Vector3 = new Vector3(1, 1, 1);
	orientation: Orientation = new Orientation();
	transformDirty: boolean = true;
	geometry: Geometry;

	protected transform: Matrix4x4 = new Matrix4x4();
	protected engine: Engine;

	constructor(engine: Engine) {
		this.engine = engine;
	}

	recalculateTransform() {
		this.transformDirty = false;

		this.transform.setScale(
			this.scale.x,
			this.scale.y,
			this.scale.z
		);

		this.transform.translate(
			this.position.x,
			this.position.y,
			this.position.z
		);

		this.transform.multiplyMatrix(this.orientation.getMatrix());
	}

	getTransformMatrix() {
		if (this.transformDirty) {
			this.recalculateTransform();
		}

		return this.transform;
	}

	draw() {
		let gl = this.engine.gl;
		let program = this.engine.currentProgram;

		this.geometry.bind(this.engine);

		gl.uniformMatrix4fv(program.uniforms["transformModel"], false, this.getTransformMatrix().array);
		gl.drawArrays(gl.TRIANGLES, 0, this.geometry.vertexCount);
	}
}
