// @client/Engine.ts
import Shim from "./Shim";
import Context from "./graphics/Context";
import Timer from "./Timer";
import Broadcaster from "@common/Broadcaster";
import SequentialBroadcaster from "@common/SequentialBroadcaster";
import InputManager from "./input/InputManager";

let coolStyle = `
background-color: black;
color: white;
font-size: 20pt;
line-height: 2;
border-radius: 8px;
padding: 4px 8px;
`;

export default class Engine {
	timer: Timer;
	context: Context;
	gl: WebGLRenderingContext;
	canvas: HTMLCanvasElement;

	onLoad = new SequentialBroadcaster();
	onUpdate = new Broadcaster();
	onDraw = new Broadcaster();

	private boundLoop: Function;

	constructor() {
		this.boundLoop = this.loop.bind(this);
	}

	logAwesome(msg: string) {
		return console.log(`%c${msg}`, coolStyle);
	}

	main() {
		if (document.readyState == "complete" ||
			document.readyState == "loaded" ||
			document.readyState == "interactive") {

			this.start();
		} else {
			document.addEventListener("DOMContentLoaded", () => {
				this.start();
			});
		}
	}

	start() {
		this.init();
		this.onLoad.fire([]).then(() => this.loop());
	}

	init() {
		this.initTimer();
		this.initCanvas();
		this.initContext();
		this.initInput();
	}

	loop() {
		this.step();
		this.draw();

		InputManager.update();

		requestAnimationFrame(<FrameRequestCallback>this.boundLoop);
	}

	draw() {
		this.onDraw.fire([]);
	}

	step() {
		this.timer.step();

		if (this.timer.delta() > 0) {
			this.onUpdate.fire([this.timer.delta()]);
		}
	}

	initInput() {
		InputManager.hook();
	}

	initTimer() {
		this.timer = new Timer();
	}

	initCanvas() {
		let canvas = document.createElement("canvas");
		this.canvas = canvas;
		canvas.width = window.innerWidth;
		canvas.height = window.innerHeight;

		canvas.addEventListener("click", (e) => {
			Shim.lockPointer(canvas);
		});

		document.body.appendChild(canvas);
	}

	initContext() {
		this.context = new Context(this.canvas);
		this.gl = this.context.gl;

		this.gl.viewport(0, 0, this.canvas.width, this.canvas.height);
	}
}
