// @client/Engine.ts
import Shim from "./Shim";
import Context from "./graphics/Context";
import Timer from "./Timer";
import Broadcaster from "@common/Broadcaster";
import SerialBroadcaster from "@common/SerialBroadcaster";
import InputManager from "./input/InputManager";
import ShaderProgram from "./graphics/ShaderProgram";

const requiredExtensions = [
	"OES_vertex_array_object",
	"WEBGL_draw_buffers",
	"OES_texture_float",
	"OES_texture_float_linear",
	"WEBGL_depth_texture"
];

const coolStyle = `
background-color: black;
color: white;
font-size: 20pt;
line-height: 2;
border-radius: 8px;
padding: 4px 8px;
`;

export default class Engine {
	debug: boolean = false;

	timer: Timer;
	context: Context;
	gl: WebGLRenderingContext;
	vao: OES_vertex_array_object;
	canvas: HTMLCanvasElement;

	onLoad = new SerialBroadcaster();
	onUpdate = new Broadcaster();
	onDraw = new Broadcaster();

	currentProgram: ShaderProgram;

	private boundLoop: Function;
	private extensions: any[];

	constructor(debug: boolean = false) {
		this.debug = debug;
		this.boundLoop = this.loop.bind(this);
	}

	useProgram(program: ShaderProgram) {
		if (program) {
			program.use();
		} else {
			this.gl.useProgram(null);
			this.currentProgram = undefined;
		}
	}

	logAwesome(msg: string) {
		return console.log(`%c${msg}`, coolStyle);
	}

	main() {
		if (document.readyState == "complete" ||
			document.readyState == "loaded" ||
			document.readyState == "interactive") {

			this.start();
		} else {
			document.addEventListener("DOMContentLoaded", () => {
				this.start();
			});
		}
	}

	start() {
		this.init();
		this.onLoad.fire([]).then(() => this.loop());
	}

	loop() {
		this.step();
		this.draw();

		InputManager.update();

		requestAnimationFrame(<FrameRequestCallback>this.boundLoop);
	}

	draw() {
		this.onDraw.fire([]);
	}

	step() {
		this.timer.step();

		if (this.timer.delta() > 0) {
			this.onUpdate.fire([this.timer.delta()]);
		}
	}

	resize(width: number, height: number) {
		this.canvas.width = width;
		this.canvas.height = height;

		this.gl.viewport(0, 0, width, height);
	}

	init() {
		this.initTimer();
		this.initCanvas();
		this.initContext();
		this.initInput();
	}

	initInput() {
		InputManager.hook();
	}

	initTimer() {
		this.timer = new Timer();
	}

	initCanvas(width: number = window.innerWidth, height: number = window.innerHeight) {
		let canvas = document.createElement("canvas");
		this.canvas = canvas;
		canvas.width = width;
		canvas.height = height;

		canvas.addEventListener("click", (e) => {
			Shim.lockPointer(canvas);
		});

		document.body.appendChild(canvas);
	}

	initContext() {
		this.context = new Context(this.canvas);

		this.gl = this.context.gl;

		this.gl.viewport(0, 0, this.canvas.width, this.canvas.height);

		let extensions = requiredExtensions.map(this.gl.getExtension.bind(this.gl));
		this.extensions = extensions;

		if (extensions.indexOf(null) > -1) {
			extensions.forEach((value, key) => {
				if (!value) {
					let name = requiredExtensions[key];
					console.error(`Couldn't load required WebGL extension ${name}!`);
				}
			});

			throw new Error("Failed to acquire some WebGL extensions.");
		}

		this.vao = this.gl.getExtension("OES_vertex_array_object");
	}
}
